### PR TITLE
Import `halo2_gadgets` as extracted from `orchard`

### PR DESCRIPTION
**Pull request by str4d**
The steps I took to produce this:
- Created a branch starting from https://github.com/zcash/orchard/commit/159ab53da59ba762794862dc4569c0273c67ed1c.
- Prepared the `halo2_gadgets` crate by:
  - Deleting Orchard-specific content.
  - Moving files into a `halo2_gadgets/` subdirectory (or otherwise ensuring they wouldn't collide with files in this repo).
- Created this branch starting from https://github.com/zcash/halo2/commit/bb56139414332da8964f92c4d5332af5848fe706.
- Added my local `zcash/orchard` repo as a git remote.
- `git merge --allow-unrelated-histories orchard/gadgets-extraction`
  - This brings the history of `zcash/orchard` into this repository, in a way that allows git history to function intuitively.
- Performed post-merge integration of `halo2_gadgets` into the workspace.

Best reviewed by-commit, starting from c2866beb4781790f1cf6c644b8b632bdd8ab843e for obvious reasons. The relevant diff is https://github.com/zcash/halo2/pull/473/files/159ab53da59ba762794862dc4569c0273c67ed1c..5312343e6d68b1bdd141337b52a5b73cce752bff

Part of zcash/orchard#183.

### CODE REVIEW

File: halo2_gadgets/src/sinsemilla/merkle.rs [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+use pasta_curves::arithmetic::CurveAffine;
+
+use super::{HashDomains, SinsemillaInstructions};
+
+use crate::utilities::{
+    cond_swap::CondSwapInstructions, i2lebsp, transpose_option_array, UtilitiesInstructions,
+};
+use std::iter;
+
+pub(crate) mod chip;
```
Comment (therealyingtong) [inline halo2_gadgets/src/sinsemilla/merkle.rs:16]:
```suggestion
pub mod chip;
```

File: halo2_gadgets/src/utilities.rs [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+use halo2_proofs::{
+    circuit::{AssignedCell, Cell, Layouter},
+    plonk::{Advice, Column, Error, Expression},
+};
+use pasta_curves::arithmetic::FieldExt;
+use std::{array, ops::Range};
+
+pub(crate) mod cond_swap;
+pub(crate) mod decompose_running_sum;
+pub(crate) mod lookup_range_check;
```
Comment (therealyingtong) [inline halo2_gadgets/src/utilities.rs:13]:
```suggestion
pub mod cond_swap;
pub mod decompose_running_sum;
pub mod lookup_range_check;
```

File: halo2_gadgets/src/sinsemilla/merkle/chip.rs [hunk 3]
```diff
@@ -0,0 +1,492 @@
+use halo2_proofs::{
```
Comment (therealyingtong) [inline halo2_gadgets/src/sinsemilla/merkle/chip.rs:1]:
```suggestion
//! Chip implementing a Merkle hash using Sinsemilla as the hash function.

use halo2_proofs::{
```

File: halo2_gadgets/src/sinsemilla/merkle/chip.rs [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+            bitrange_subset,
+            cond_swap::{CondSwapChip, CondSwapConfig, CondSwapInstructions},
+            UtilitiesInstructions,
+        },
+    },
+};
+use group::ff::PrimeField;
+use std::array;
+
+#[derive(Clone, Debug)]
```
Comment (therealyingtong) [inline halo2_gadgets/src/sinsemilla/merkle/chip.rs:28]:
```suggestion
/// Configuration for the `MerkleChip` implementation.
#[derive(Clone, Debug)]
```

File: halo2_gadgets/src/sinsemilla/merkle/chip.rs [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+    Fixed: FixedPoints<pallas::Affine>,
+    Commit: CommitDomains<pallas::Affine, Fixed, Hash>,
+{
+    advices: [Column<Advice>; 5],
+    q_decompose: Selector,
+    pub(super) cond_swap_config: CondSwapConfig,
+    pub(super) sinsemilla_config: SinsemillaConfig<Hash, Commit, Fixed>,
+}
+
+#[derive(Clone, Debug)]
```
Comment (therealyingtong) [inline halo2_gadgets/src/sinsemilla/merkle/chip.rs:41]:
```suggestion
/// Chip implementing `MerkleInstructions`.
#[derive(Clone, Debug)]
```

File: halo2_gadgets/src/sinsemilla/merkle/chip.rs [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+    }
+}
+
+impl<Hash, Commit, F> MerkleChip<Hash, Commit, F>
+where
+    Hash: HashDomains<pallas::Affine>,
+    F: FixedPoints<pallas::Affine>,
+    Commit: CommitDomains<pallas::Affine, F, Hash>,
+{
+    pub fn configure(
```
Comment (therealyingtong) [inline halo2_gadgets/src/sinsemilla/merkle/chip.rs:75]:
```suggestion
    /// Configures the [`MerkleChip`].
    pub fn configure(
```

File: halo2_gadgets/src/sinsemilla/merkle/chip.rs [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+
+        MerkleConfig {
+            advices,
+            q_decompose,
+            cond_swap_config,
+            sinsemilla_config,
+        }
+    }
+
+    pub fn construct(config: MerkleConfig<Hash, Commit, F>) -> Self {
```
Comment (therealyingtong) [inline halo2_gadgets/src/sinsemilla/merkle/chip.rs:178]:
```suggestion
    /// Constructs a [`MerkleChip`] given a [`MerkleConfig`].
    pub fn construct(config: MerkleConfig<Hash, Commit, F>) -> Self {
```

File: halo2_gadgets/src/utilities/cond_swap.rs [hunk 8]
```diff
@@ -0,0 +1,294 @@
+use super::{bool_check, ternary, UtilitiesInstructions};
```
Comment (therealyingtong) [inline halo2_gadgets/src/utilities/cond_swap.rs:1]:
```suggestion
//! Gadget and chip for a conditional swap utility.

use super::{bool_check, ternary, UtilitiesInstructions};
```

File: halo2_gadgets/src/utilities/cond_swap.rs [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+use halo2_proofs::{
+    circuit::{AssignedCell, Chip, Layouter},
+    plonk::{Advice, Column, ConstraintSystem, Error, Selector},
+    poly::Rotation,
+};
+use pasta_curves::arithmetic::FieldExt;
+use std::{array, marker::PhantomData};
+
+pub trait CondSwapInstructions<F: FieldExt>: UtilitiesInstructions<F> {
```
Comment (therealyingtong) [inline halo2_gadgets/src/utilities/cond_swap.rs:10]:
```suggestion
/// Instructions for a conditional swap gadget.
pub trait CondSwapInstructions<F: FieldExt>: UtilitiesInstructions<F> {
```

File: halo2_gadgets/src/utilities/cond_swap.rs [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+        &self.config
+    }
+
+    fn loaded(&self) -> &Self::Loaded {
+        &()
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct CondSwapConfig {
```
Comment (therealyingtong) [inline halo2_gadgets/src/utilities/cond_swap.rs:46]:
```suggestion
/// Configuration for the [`CondSwapChip`].
#[derive(Clone, Debug)]
pub struct CondSwapConfig {
```

File: halo2_gadgets/src/utilities/cond_swap.rs [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+            let bool_check = bool_check(swap);
+
+            array::IntoIter::new([a_check, b_check, bool_check])
+                .map(move |poly| q_swap.clone() * poly)
+        });
+
+        config
+    }
+
+    pub fn construct(config: CondSwapConfig) -> Self {
```
Comment (therealyingtong) [inline halo2_gadgets/src/utilities/cond_swap.rs:188]:
```suggestion
    /// Constructs a [`CondSwapChip`] given a [`CondSwapConfig`].
    pub fn construct(config: CondSwapConfig) -> Self {
```

File: halo2_gadgets/src/utilities/cond_swap.rs [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+}
+
+#[derive(Clone, Debug)]
+pub struct CondSwapConfig {
+    pub q_swap: Selector,
+    pub a: Column<Advice>,
+    pub b: Column<Advice>,
+    pub a_swapped: Column<Advice>,
+    pub b_swapped: Column<Advice>,
+    pub swap: Column<Advice>,
```
Comment (therealyingtong) [inline halo2_gadgets/src/utilities/cond_swap.rs:52]:
```suggestion
    q_swap: Selector,
    a: Column<Advice>,
    b: Column<Advice>,
    a_swapped: Column<Advice>,
    b_swapped: Column<Advice>,
    swap: Column<Advice>,
```

File: halo2_gadgets/src/utilities/cond_swap.rs [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+}
+
```
Comment (therealyingtong) [inline halo2_gadgets/src/utilities/cond_swap.rs:54]:
```suggestion

#[cfg(test)]
impl CondSwapConfig {
    pub(crate) fn a(&self) -> Column<Advice> {
        self.a
    }
}

```

File: halo2_gadgets/src/utilities/decompose_running_sum.rs [hunk 14]
```diff
@@ -0,0 +1,391 @@
...
+    plonk::{Advice, Column, ConstraintSystem, Error, Selector},
+    poly::Rotation,
+};
+
+use super::range_check;
+use pasta_curves::arithmetic::FieldExt;
+use std::marker::PhantomData;
+
+/// The running sum $[z_0, ..., z_W]$. If created in strict mode, $z_W = 0$.
+pub struct RunningSum<F: FieldExt + PrimeFieldBits>(Vec<AssignedCell<F, F>>);
```
Comment (therealyingtong) [inline halo2_gadgets/src/utilities/decompose_running_sum.rs:37]:
```suggestion
/// The running sum $[z_0, ..., z_W]$. If created in strict mode, $z_W = 0$.
#[derive(Debug)]
pub struct RunningSum<F: FieldExt + PrimeFieldBits>(Vec<AssignedCell<F, F>>);
```

File: halo2_gadgets/src/utilities/decompose_running_sum.rs [hunk 15]
```diff
@@ -0,0 +1,391 @@
...
+        &self.0
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub struct RunningSumConfig<F: FieldExt + PrimeFieldBits, const WINDOW_NUM_BITS: usize> {
+    pub q_range_check: Selector,
+    pub z: Column<Advice>,
+    _marker: PhantomData<F>,
+}
```
Comment (therealyingtong) [inline halo2_gadgets/src/utilities/decompose_running_sum.rs:51]:
```suggestion
/// Configuration that provides methods for running sum decomposition.
#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub struct RunningSumConfig<F: FieldExt + PrimeFieldBits, const WINDOW_NUM_BITS: usize> {
    q_range_check: Selector,
    z: Column<Advice>,
    _marker: PhantomData<F>,
}
```

File: halo2_gadgets/src/utilities/decompose_running_sum.rs [hunk 16]
```diff
@@ -0,0 +1,391 @@
...
+
+impl<F: FieldExt + PrimeFieldBits, const WINDOW_NUM_BITS: usize>
+    RunningSumConfig<F, WINDOW_NUM_BITS>
+{
```
Comment (therealyingtong) [inline halo2_gadgets/src/utilities/decompose_running_sum.rs:55]:
```suggestion
impl<F: FieldExt + PrimeFieldBits, const WINDOW_NUM_BITS: usize>
    RunningSumConfig<F, WINDOW_NUM_BITS>
{
    /// Returns the q_range_check selector of this [`RunningSumConfig`].
    pub fn q_range_check(&self) -> Selector {
        self.q_range_check
    }

```

File: halo2_gadgets/src/ecc/chip/mul_fixed.rs [hunk 17]
```diff
@@ -0,0 +1,505 @@
...
+    /// This gate is used both in the mul_fixed::base_field_elem and mul_fixed::short
+    /// helpers, which decompose the scalar using a running sum.
+    ///
+    /// This gate is not used in the mul_fixed::full_width helper, since the full-width
+    /// scalar is witnessed directly as three-bit windows instead of being decomposed
+    /// via a running sum.
+    fn running_sum_coords_gate(&self, meta: &mut ConstraintSystem<pallas::Base>) {
+        meta.create_gate("Running sum coordinates check", |meta| {
+            let q_mul_fixed_running_sum =
+                meta.query_selector(self.running_sum_config.q_range_check);
```
Comment (therealyingtong) [inline halo2_gadgets/src/ecc/chip/mul_fixed.rs:133]:
```suggestion
                meta.query_selector(self.running_sum_config.q_range_check());
```

File: halo2_gadgets/src/ecc/chip/mul_fixed/base_field_elem.rs [hunk 18]
```diff
@@ -0,0 +1,517 @@
...
+                        running_sum: (*running_sum).as_slice().try_into().unwrap(),
+                    }
+                };
+
+                let (acc, mul_b) = self.super_config.assign_region_inner::<_, NUM_WINDOWS>(
+                    &mut region,
+                    offset,
+                    &(&scalar).into(),
+                    base,
+                    self.super_config.running_sum_config.q_range_check,
```
Comment (therealyingtong) [inline halo2_gadgets/src/ecc/chip/mul_fixed/base_field_elem.rs:193]:
```suggestion
                    self.super_config.running_sum_config.q_range_check(),
```

File: halo2_gadgets/src/ecc/chip/mul_fixed/short.rs [hunk 19]
```diff
@@ -0,0 +1,617 @@
...
+                let scalar = self.decompose(&mut region, offset, magnitude_sign.clone())?;
+
+                let (acc, mul_b) = self
+                    .super_config
+                    .assign_region_inner::<_, NUM_WINDOWS_SHORT>(
+                        &mut region,
+                        offset,
+                        &(&scalar).into(),
+                        base,
+                        self.super_config.running_sum_config.q_range_check,
```
Comment (therealyingtong) [inline halo2_gadgets/src/ecc/chip/mul_fixed/short.rs:120]:
```suggestion
                        self.super_config.running_sum_config.q_range_check(),
```

File: halo2_gadgets/src/utilities/lookup_range_check.rs [hunk 20]
```diff
@@ -0,0 +1,629 @@
...
+impl<F: FieldExt + PrimeFieldBits> std::ops::Deref for RunningSum<F> {
+    type Target = Vec<AssignedCell<F, F>>;
+
+    fn deref(&self) -> &Vec<AssignedCell<F, F>> {
+        &self.0
+    }
+}
+
+#[derive(Eq, PartialEq, Debug, Clone, Copy)]
+pub struct LookupRangeCheckConfig<F: FieldExt + PrimeFieldBits, const K: usize> {
```
Comment (therealyingtong) [inline halo2_gadgets/src/utilities/lookup_range_check.rs:27]:
```suggestion
/// Configuration that provides methods for a lookup range check.
#[derive(Eq, PartialEq, Debug, Clone, Copy)]
pub struct LookupRangeCheckConfig<F: FieldExt + PrimeFieldBits, const K: usize> {
```

File: halo2_gadgets/src/utilities/lookup_range_check.rs [hunk 21]
```diff
@@ -0,0 +1,629 @@
...
+    pub q_lookup: Selector,
+    pub q_running: Selector,
+    pub q_bitshift: Selector,
+    pub running_sum: Column<Advice>,
+    table_idx: TableColumn,
+    _marker: PhantomData<F>,
+}
```
Comment (therealyingtong) [inline halo2_gadgets/src/utilities/lookup_range_check.rs:34]:
```suggestion
/// Configuration that provides methods for a lookup range check.
#[derive(Eq, PartialEq, Debug, Clone, Copy)]
pub struct LookupRangeCheckConfig<F: FieldExt + PrimeFieldBits, const K: usize> {
    q_lookup: Selector,
    q_running: Selector,
    q_bitshift: Selector,
    running_sum: Column<Advice>,
    table_idx: TableColumn,
    _marker: PhantomData<F>,
}
```

File: halo2_gadgets/src/sinsemilla/merkle.rs [hunk 22]
```diff
@@ -0,0 +1,356 @@
...
+        MerkleChip,
+        const PATH_LENGTH: usize,
+        const K: usize,
+        const MAX_WORDS: usize,
+    > MerklePath<C, MerkleChip, PATH_LENGTH, K, MAX_WORDS>
+where
+    MerkleChip: MerkleInstructions<C, PATH_LENGTH, K, MAX_WORDS> + Clone,
+{
+    /// Calculates the root of the tree containing the given leaf at this Merkle path.
+    pub(crate) fn calculate_root(
```
Comment (therealyingtong) [inline halo2_gadgets/src/sinsemilla/merkle.rs:81]:
```suggestion
    pub fn calculate_root(
```

File: halo2_gadgets/src/sinsemilla/merkle.rs [hunk 23]
```diff
@@ -0,0 +1,356 @@
...
+> where
+    MerkleChip: MerkleInstructions<C, PATH_LENGTH, K, MAX_WORDS> + Clone,
+{
+    pub(crate) chip_1: MerkleChip,
+    pub(crate) chip_2: MerkleChip,
+    pub(crate) domain: MerkleChip::HashDomains,
+    pub(crate) leaf_pos: Option<u32>,
+    // The Merkle path is ordered from leaves to root.
+    pub(crate) path: Option<[C::Base; PATH_LENGTH]>,
+}
```
Comment (therealyingtong) [inline halo2_gadgets/src/sinsemilla/merkle.rs:67]:
```suggestion
pub struct MerklePath<
    C: CurveAffine,
    MerkleChip,
    const PATH_LENGTH: usize,
    const K: usize,
    const MAX_WORDS: usize,
> where
    MerkleChip: MerkleInstructions<C, PATH_LENGTH, K, MAX_WORDS> + Clone,
{
    chip_1: MerkleChip,
    chip_2: MerkleChip,
    domain: MerkleChip::HashDomains,
    leaf_pos: Option<u32>,
    // The Merkle path is ordered from leaves to root.
    path: Option<[C::Base; PATH_LENGTH]>,
}

impl<
        C: CurveAffine,
        MerkleChip,
        const PATH_LENGTH: usize,
        const K: usize,
        const MAX_WORDS: usize,
    > MerklePath<C, MerkleChip, PATH_LENGTH, K, MAX_WORDS>
where
    MerkleChip: MerkleInstructions<C, PATH_LENGTH, K, MAX_WORDS> + Clone,
{
    fn construct(
        chip_1: MerkleChip,
        chip_2: MerkleChip,
        domain: MerkleChip::HashDomains,
        leaf_pos: Option<u32>,
        path: Option<[C::Base; PATH_LENGTH]>,
    ) -> Self {
        Self {
            chip_1,
            chip_2,
            domain,
            leaf_pos,
            path,
        }
    }
}
```

**Review by therealyingtong — CHANGES_REQUESTED** [review]
I made these changes https://github.com/zcash/halo2/pull/475 in order to get it working in the `orchard` crate: https://github.com/zcash/orchard/pull/217.

File: halo2_gadgets/Cargo.toml [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+
+[target.'cfg(unix)'.dev-dependencies]
+pprof = { version = "=0.6.1", features = ["criterion", "flamegraph"] }
+
+[lib]
+bench = false
+
+[features]
+dev-graph = ["halo2_proofs/dev-graph", "plotters"]
+test-dependencies = ["proptest"]
```
Comment (therealyingtong) [inline halo2_gadgets/Cargo.toml:51]: Are we using `test-dependencies` anywhere?

Comment (str4d) [reply halo2_gadgets/Cargo.toml]: Nope, it was left over from `orchard`. I've removed it.

**Review by therealyingtong — CHANGES_REQUESTED** [review]
More visibility fixes:  https://github.com/zcash/halo2/pull/476

**Review by therealyingtong — APPROVED** [review]
utACK. We can remove Orchard-specific documentation in a future PR.

**Review by ebfull — APPROVED** [review]

**Pull request merged**